### PR TITLE
Fix #4006: Avoid ambiguity between SHAREDterm cases and finally

### LINKED ISSUE #4006: Inlined try block breaks pickling tests
```scala
class Foo {
  inline def foo: Int = try { 1 } finally println("Hello")
  foo
}
```

```
$ dotc tests/allan/Test.scala -Ytest-pickler 
exception occurred while compiling tests/allan/Test.scala
Exception in thread "main" class dotty.tools.dotc.reporting.diagnostic.messages$Error at ?: pickling difference for class Foo in tests/allan/Test.scala, for details:
                   |
                   |  diff before-pickling.txt after-pickling.txt
	at dotty.tools.dotc.reporting.Reporting.error(Reporter.scala:88)
	at dotty.tools.dotc.transform.Pickler.testSame(Pickler.scala:107)
	at dotty.tools.dotc.transform.Pickler.testUnpickler$$anonfun$4(Pickler.scala:99)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
	at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:138)
	at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:236)
	at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:229)
	at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
	at scala.collection.mutable.HashMap.foreach(HashMap.scala:138)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:788)
	at dotty.tools.dotc.transform.Pickler.testUnpickler(Pickler.scala:100)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:83)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:123)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:18)
	at scala.compat.java8.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:136)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:141)
	at scala.compat.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:88)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:143)
	at dotty.tools.dotc.Run.compileSources(Run.scala:94)
	at dotty.tools.dotc.Run.compile(Run.scala:78)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:29)
	at dotty.tools.dotc.Driver.process(Driver.scala:127)
	at dotty.tools.dotc.Driver.process(Driver.scala:96)
	at dotty.tools.dotc.Driver.process(Driver.scala:108)
	at dotty.tools.dotc.Driver.main(Driver.scala:135)
	at dotty.tools.dotc.Main.main(Main.scala
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/test/dotty/tools/dotc/CompilationTests.scala [hunk 1]
```diff
@@ -58,6 +58,7 @@ class CompilationTests extends ParallelTesting {
     compileFile("tests/pos-special/i3323.scala", defaultOptions.and("-Xfatal-warnings")) +
     compileFile("tests/pos-special/i3323b.scala", defaultOptions.and("-Xfatal-warnings")) +
     compileFile("tests/pos-special/i3589-b.scala", defaultOptions.and("-Xfatal-warnings")) +
+    compileFile("tests/pos-special/i4006.scala", defaultOptions.and("-Ytest-pickler")) +
```
Comment (allanrenucci) [inline compiler/test/dotty/tools/dotc/CompilationTests.scala]: Why not put the test in `tests/pickling`?

**Comment by nicolasstucki** [comment]
Added description of the issue in #4006

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 2]
```diff
@@ -423,7 +423,11 @@ class TreePickler(pickler: TastyPickler) {
           withLength { pickleSymRef(from.symbol); pickleTreeUnlessEmpty(expr) }
         case Try(block, cases, finalizer) =>
           writeByte(TRY)
-          withLength { pickleTree(block); cases.foreach(pickleTree); pickleTreeUnlessEmpty(finalizer) }
+          withLength {
+            pickleTree(block)
+            cases.foreach(pickleTree)
+            if (!finalizer.isEmpty) writeByte(FINALLY); pickleTreeUnlessEmpty(finalizer)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala]:
```scala
if (!finalizer.isEmpty) {
  writeByte(FINALLY)
  pickleTree(finalizer)
}
```

File: compiler/test/dotty/tools/dotc/CompilationTests.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/dotc/CompilationTests.scala]: Yes

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala]: Will completely rewrite this code

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 3]
```diff
@@ -1021,7 +1021,12 @@ class TreeUnpickler(reader: TastyReader,
               val expr = ifBefore(end)(readTerm(), EmptyTree)
               Return(expr, Ident(from.termRef))
             case TRY =>
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: Maybe, also add a test with no `finally` to exercise this path. I found tests for `try ... catch` but not within `inline def`

**Comment by odersky** [comment]
I think we should not add a new format tag. Tasty cases are a precious resource - the more we add the more people will get the impression that this is a complex format and will not work with it. 

I suppose we can use `unsharedTag` instead to get a robust disambiguation?

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala]: Added

**Pull request merged**